x-pack/plugins/security/server/session_management/session_index.ts

```diff
@@ -484,24 +553,30 @@ export class SessionIndex {
       });
     }
 
-    try {
-      const { body: response } = await this.options.elasticsearchClient.deleteByQuery(
-        {
-          index: this.indexName,
-          refresh: true,
-          body: { query: { bool: { should: deleteQueries } } },
-        },
-        { ignore: [409, 404] }
-      );
+    // Create point in time snapshot to paginate through sessions
+    const { body: pitResponse } = await this.options.elasticsearchClient.openPointInTime({
+      index: this.indexName,
+      keep_alive: SESSION_INDEX_CLEANUP_KEEP_ALIVE,
+    });
 
-      if (response.deleted! > 0) {
-        this.options.logger.debug(
-          `Cleaned up ${response.deleted} invalid or expired session values.`
-        );
+    // We don't know the total number of sessions until we fetched the first batch so assume we have at least one session to clean.
+    let total = 1;
+    let searchAfter: SortResults | undefined;
+    for (let i = 0; i < total / SESSION_INDEX_CLEANUP_BATCH_SIZE; i++) {
+      const { body: searchResponse } =
+        await this.options.elasticsearchClient.search<SessionIndexValue>({
+          pit: { id: pitResponse.id, keep_alive: SESSION_INDEX_CLEANUP_KEEP_ALIVE },
+          _source_includes: 'usernameHash,provider',
+          query: { bool: { should: deleteQueries } },
+          search_after: searchAfter,
+          size: SESSION_INDEX_CLEANUP_BATCH_SIZE,
+          sort: '_shard_doc',
+        });
+      if (searchResponse.hits.hits.length > 0) {
+        yield searchResponse.hits.hits;
+        total = (searchResponse.hits.total as SearchTotalHits).value;
+        searchAfter = searchResponse.hits.hits[searchResponse.hits.hits.length - 1].sort;
       }
```

Comment:
I tested this and if you have > 10k sessions, this loop terminates after one iteration, because the `total` is > 10k.

I don't think you need to use the `total` here at all, in fact after looking into this a bit more, [the total hits is not accurate for amounts > 10k anyway _and_ we get a performance improvement if we disable it](https://www.elastic.co/guide/en/elasticsearch/reference/8.0/search-your-data.html#track-total-hits).

Furthermore, this loop has no limit, [as I mentioned above](https://github.com/elastic/kibana/pull/122419#discussion_r781430585) we need _some_ sort of limit.

I suggest changing the loop like so, I tested this and it worked well:

```diff
diff --git a/x-pack/plugins/security/server/session_management/session_index.ts b/x-pack/plugins/security/server/session_management/session_index.ts
index 037c47ca791..554d917b834 100644
--- a/x-pack/plugins/security/server/session_management/session_index.ts
+++ b/x-pack/plugins/security/server/session_management/session_index.ts
@@ -7,7 +7,6 @@
 
 import type {
   BulkOperationContainer,
-  SearchTotalHits,
   SortResults,
 } from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
 
@@ -48,6 +47,12 @@ const SESSION_INDEX_TEMPLATE_VERSION = 1;
  */
 const SESSION_INDEX_CLEANUP_BATCH_SIZE = 10_000;
 
+/**
+ * Maximum number of batches per cleanup.
+ * If the batch size is 10,000 and this limit is 10, then Kibana will remove up to 100k sessions per cleanup.
+ */
+const SESSION_INDEX_CLEANUP_BATCH_LIMIT = 10;
+
 /**
  * Number of sessions to remove per batch during cleanup.
  */
@@ -559,10 +564,8 @@ export class SessionIndex {
       keep_alive: SESSION_INDEX_CLEANUP_KEEP_ALIVE,
     });
 
-    // We don't know the total number of sessions until we fetched the first batch so assume we have at least one session to clean.
-    let total = 1;
     let searchAfter: SortResults | undefined;
-    for (let i = 0; i < total / SESSION_INDEX_CLEANUP_BATCH_SIZE; i++) {
+    for (let i = 0; i < SESSION_INDEX_CLEANUP_BATCH_LIMIT; i++) {
       const { body: searchResponse } =
         await this.options.elasticsearchClient.search<SessionIndexValue>({
           pit: { id: pitResponse.id, keep_alive: SESSION_INDEX_CLEANUP_KEEP_ALIVE },
@@ -571,11 +574,15 @@ export class SessionIndex {
           search_after: searchAfter,
           size: SESSION_INDEX_CLEANUP_BATCH_SIZE,
           sort: '_shard_doc',
+          track_total_hits: false, // for performance
         });
-      if (searchResponse.hits.hits.length > 0) {
-        yield searchResponse.hits.hits;
-        total = (searchResponse.hits.total as SearchTotalHits).value;
-        searchAfter = searchResponse.hits.hits[searchResponse.hits.hits.length - 1].sort;
+      const { hits } = searchResponse.hits;
+      if (hits.length > 0) {
+        yield hits;
+        searchAfter = hits[hits.length - 1].sort;
+        if (hits.length < SESSION_INDEX_CLEANUP_BATCH_SIZE) {
+          break;
+        }
       }
     }
   }
```

**In addition**: can you please add two unit tests -- one for 10,001 sessions to demonstrate the loop terminates early, and one for 100,001 sessions to demonstrate that the loop doesn't exceed the configured limit.